pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java

```diff
@@ -4,262 +4,52 @@
-
-        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
-            n = (JavaNode) node.jjtGetChild(i);
-            npath += (Integer) n.jjtAccept(this, data);
-        }
-
-        return npath;
-    }
-
-    @Override
-    public Object visit(ASTMethodDeclaration node, Object data) {
-        int npath = complexityMultipleOf(node, 1, data);
-
-        DataPoint point = new DataPoint();
-        point.setNode(node);
-        point.setScore(1.0 * npath);
-        point.setMessage(getMessage());
-        addDataPoint(point);
-
-        return Integer.valueOf(npath);
-    }
-
-    @Override
-    public Object visit(JavaNode node, Object data) {
-        int npath = complexityMultipleOf(node, 1, data);
-        return Integer.valueOf(npath);
-    }
-
-    @Override
-    public Object visit(ASTIfStatement node, Object data) {
-        // (npath of if + npath of else (or 1) + bool_comp of if) * npath of
-        // next
-
-        List<JavaNode> statementChildren = new ArrayList<>();
-        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
-            if (node.jjtGetChild(i).getClass() == ASTStatement.class) {
-                statementChildren.add((JavaNode) node.jjtGetChild(i));
-            }
-        }
-
-        if (statementChildren.isEmpty() || statementChildren.size() == 1 && node.hasElse()
-                || statementChildren.size() != 1 && !node.hasElse()) {
-            throw new IllegalStateException("If node has wrong number of children");
-        }
-
-        // add path for not taking if
-        int complexity = 0;
-        if (!node.hasElse()) {
-            complexity++;
-        }
-
-        for (JavaNode element : statementChildren) {
-            complexity += (Integer) element.jjtAccept(this, data);
-        }
-
-        int boolCompIf = sumExpressionComplexity(node.getFirstChildOfType(ASTExpression.class));
-        return Integer.valueOf(boolCompIf + complexity);
-    }
-
-    @Override
-    public Object visit(ASTWhileStatement node, Object data) {
-        // (npath of while + bool_comp of while + 1) * npath of next
-
-        int boolCompWhile = sumExpressionComplexity(node.getFirstChildOfType(ASTExpression.class));
-
-        Integer nPathWhile = (Integer) ((JavaNode) node.getFirstChildOfType(ASTStatement.class)).jjtAccept(this, data);
+public class NPathComplexityRule extends AbstractJavaMetricsRule {
 
-        return Integer.valueOf(boolCompWhile + nPathWhile + 1);
-    }
-
-    @Override
-    public Object visit(ASTDoStatement node, Object data) {
-        // (npath of do + bool_comp of do + 1) * npath of next
-
-        int boolCompDo = sumExpressionComplexity(node.getFirstChildOfType(ASTExpression.class));
-
-        Integer nPathDo = (Integer) ((JavaNode) node.getFirstChildOfType(ASTStatement.class)).jjtAccept(this, data);
+    private static final IntegerProperty REPORT_LEVEL_DESCRIPTOR = new IntegerProperty(
+        "reportLevel", "N-Path Complexity reporting threshold", 1, 30, 200, 1.0f);
```

Comment:
If I'm reading this correctly, the property has a minimum value of 1, a maximum of 30, and a default of 200??

Something doesn't sound right about that...

Reply:
Hmm curiously the descriptor doesn't fail during construction... We should add a test case to the numeric descriptors and correct that

Reply:
I added #725 to track that

Reply:
Btw: the original MINIMUM_DESCRIPTOR had also a range 0..100, and we used 200 as default :)